File: day10/go/part2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	f, err := os.Open("../../input")
	if err != nil {
		panic(err)
	}

	s := bufio.NewScanner(f)

	const width, height = 40, 6
	var addxV *int
	x := 1

	for cycle := 1; cycle <= width*height; cycle++ {
		rayX := (cycle - 1) % width
		if x+1 >= rayX && rayX >= x-1 {
			os.Stdout.Write([]byte{'#'})
		} else {
			os.Stdout.Write([]byte{'.'})
		}
		if rayX == width-1 {
			fmt.Println()
		}

		if addxV != nil {
			x += *addxV
			addxV = nil
			continue
		}

		if !s.Scan() {
			break
		}
		line := s.Text()
		opEnd := strings.IndexByte(line, ' ')
		if opEnd == -1 {
			opEnd = len(line)
		}

		switch line[:opEnd] {
		case "addx":
			i, err := strconv.Atoi(line[opEnd+1:])
			if err != nil {
				panic(err)
			}
			addxV = &i
		case "noop":
		default:
			panic("unexpected instruction")
		}
	}
	if s.Err() != nil {
		panic(err)
	}
}

File: day10/nim/part2/main.nim
import std/options
from std/strutils import parseInt

let f = open("../../input")

const
  width = 40
  height = 6

var
  addxVal = none(int)
  x = 1

for cycle in 1..(width * height):
  let rayX = (cycle - 1) mod width
  stdout.write(
    if abs(rayX - x) <= 1: '#'
    else: '.'
  )
  if rayX == width - 1:
    stdout.write('\n')

  if addxVal.isSome:
    x += addxVal.get()
    addxVal = none(int)
    continue

  var line: string
  line = f.readLine()

  var space = line.find(' ')
  if space == -1:
    space = len(line)
  case line[0..space-1]
  of "addx":
    addxVal = some(parseInt(line[space + 1..len(line) - 1]))
  of "noop": discard
  else:
    raise newException(ValueError, "unexpected instruction")
